Pointer communicator

RiccardoRossi opened the pull request:
this provides the framework for gathering generic data by the use of global pointers.

It takes ideas of PGAS (Partitioned Global Addres Space)  in providing a proxy to remote results.

@jcotela the "generic" serializer-based SendRecv funciton is implemented in the PointerCommunicator class. As agreed it should be moved to the data communicator.

RiccardoRossi committed:
proposal for access to remote data by function

RiccardoRossi committed:
small correction

RiccardoRossi committed:
making use of the generic parallel environment

RiccardoRossi committed:
making public the default constructor. Essential to be able to work with maps :-(

RiccardoRossi committed:
utility to retrieve non local pointers for indexed objects

RiccardoRossi committed:
utility to communicate data indexing by global pointers

RiccardoRossi committed:
testing the pointer_comm

RiccardoRossi committed:
adding result proxy

RiccardoRossi committed:
formatting and avoiding work in the non-distributed case

philbucher commented:
this seems like a cool feature :+1: 

You are using synchronous communication right?
Wouldn't it be smarter/more efficient to do this asynchronous?
=> I can imagine this to be used a lot in the future, and then refactoring will be painful

When I rewrote th MappingApp not so long ago I switched from sync to async and found that the implementation was even simpler

What do you think?

RiccardoRossi commented:
@philbucher there is surely space for switching to async communications inside, nevertheless this is transparent to the user so i would leave it for later.

jcotela commented on kratos/mpi/utilities/pointer_communicator.h:
```diff
@@ -0,0 +1,373 @@
+
+namespace Kratos
+{
+///@addtogroup ApplicationNameApplication
```
Please update or remove

RiccardoRossi replied:
done

jcotela commented on kratos/utilities/global_pointer_utilities.h:
```diff
@@ -0,0 +1,465 @@
+                }
+            }
+
+            // very useful for debugging. do not remove for now
```
"for now" :man_facepalming:

jcotela commented on kratos/utilities/global_pointer_utilities.h:
```diff
@@ -0,0 +1,465 @@
+    ///@}
+    ///@name Member Variables
+    ///@{
+    DataCommunicator& mrDataCommunicator;
```
Most likely can be `const`

RiccardoRossi replied:
done

jcotela commented on kratos/mpi/utilities/pointer_communicator.h:
```diff
@@ -0,0 +1,373 @@
+     * and retrieves the result to the caller rank
+     */
+    template< class TSendType >
+    ResultsProxy<TPointerDataType, TSendType> Apply(std::function< TSendType(GlobalPointer<TPointerDataType>&) > user_function)
```
Can/should the argument of the function to be applied remotely be const (`const GlobalPointer<TPointerDataType>&` or maybe `GlobalPointer<const TPointerDataType>&`?). In either case, I assume that we do not want to allow the global pointer operation to modify the remote object.

RiccardoRossi replied:
why not? imagine you want to mark a flag as visited...

jcotela commented on kratos/mpi/utilities/pointer_communicator.h:
```diff
@@ -0,0 +1,373 @@
+private:
+    const int mcurrent_rank;
+    GlobalPointersUnorderedMap< TPointerDataType, TSendType > mNonLocalData;
+    std::function< TSendType(GlobalPointer<TPointerDataType>&) > muser_function;
```
The format here is inconsistent. I assume it should be `mCurrentRank`, `mNonLocalData`, `mUserFunction`.

RiccardoRossi replied:
done

jcotela reviewed (commented):
Overall I am positive about this PR. How should we proceed regarding the things that need to be integrated into the data communicator(and eventually the wrappers for separate Send/Recv calls?).

philbucher commented:
@RiccardoRossi I am folowing the global pointers stuff since the beginning but I still don't get how to use them :)
Do you have an example somewhere? I think it would be quite helpful for the mpi-mapping stuff

RiccardoRossi commented:
take a look at the test example. i will try anyway to post an rxplanation
here asap

On Sat, Apr 6, 2019, 1:17 PM Philipp Bucher <notifications@github.com>
wrote:

> @RiccardoRossi <https://github.com/RiccardoRossi> I am folowing the
> global pointers stuff since the beginning but I still don't get how to use
> them :)
> Do you have an example somewhere? I think it would be quite helpful for
> the mpi-mapping stuff
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/KratosMultiphysics/Kratos/pull/4592#issuecomment-480495964>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AHr7ESd1UOHUjvpYYWe7Xa9I1uiBvkAzks5veIJCgaJpZM4ceqax>
> .
>

philbucher commented:
Thanks, I hadn't seen it.

We should speak abt this next week, I think this will make my life much easier :D

RiccardoRossi commented:
here goes a quick note about the intended use of global pointers. What is implemented right now i**s a way to simplify gathering of arbitrary data**. At a later step a function for sending data will also be implemented, nevertheless we (@KratosMultiphysics/technical-committee) we did not yet converge on a good interface so sending data is left as a second step.

so back to what we have right now.

let's consider a loop to take the average of neighbours nodes for each node.
In serial it looks something like

~~~cpp
for(auto& node : model_part.Nodes)
    double Tavg = 0;
    const auto& neighbours = node.GetValue(NODAL_NEIGHBOURS);
    for(auto& neighb : neighbours)
         Tavg += neighb.FastGetSolutionStepValue(TEMPERATURE);
    node.SetValue(AVERAGE_TEMPERATURE, Tavg/(neigbours.size());
}
~~~

this loop is obviously not possible in mpi. (we do not have all the neighbours on a given process unless we add a lot of ghost nodes)

Let's now consider we use global pointers instead of weak_pointers for the neighbours. IN THE SERIAL CASE the previous loop would look very similar, something like 

~~~cpp
for(auto& node : model_part.Nodes)
    double Tavg = 0;
    const auto& global_pointers_to_neighbours = node.GetValue(GP_NODAL_NEIGHBOURS);
    for(auto& gp : global_pointers_to_neighbours)
         Tavg += gp->FastGetSolutionStepValue(TEMPERATURE);
    node.SetValue(AVERAGE_TEMPERATURE, Tavg/(global_pointers_to_neighbours.size());
}
~~~

Since global pointers are pointers that are only valid **in the memory space of the owner processor**
this works ok on a single process, but still fails in mpi, since the global neighbours can belong to different mpi processes.

The current PR makes an attempt to simplify dealing with this sort of situations:
in order to work in mpi we would have to modify the code as follows:


~~~cpp
//first of all collect all the neighbours into a list
GlobalPointerVector< Node<3 > > gp_list;

for(auto& node : model_part.Nodes)
    for(auto& gp : global_pointers_to_neighbours)
        gp_list.push_back(gp);

//now create the pointer communicator
PointerCommunicator pointer_comm(DefaultDataCommunicator, gp_list); //here we prepare the communication

//define a function to access to the data
result_proxy = pointer_comm.Apply<double>(
            [](GlobalPointer<Node<3> >& gp) -> double
            {
                  return gp->FastGetSolutionStepValue(TEMPERATURE);
            }
      );  //here all communications happen !!

for(auto& node : model_part.Nodes)
    double Tavg = 0;
    const auto& global_pointers_to_neighbours = node.GetValue(GP_NODAL_NEIGHBOURS);
    for(auto& gp : global_pointers_to_neighbours)
         Tavg += result_proxy.Get(gp); //here we get the result of the application of the function onto gp. This is so BOTH for LOCAL GPs and for REMOTE GPs
    node.SetValue(AVERAGE_TEMPERATURE, Tavg/(global_pointers_to_neighbours.size());
}
~~~

the code is of course a little more verbose, nevertheless all the communication is hidden in the Apply function (and in the constructor), and the use if still more or less similar to the original serial code.

Aside of this, the code can be made to be still efficient in serial. The point here is that the function "get" of the "result_proxy" is actually **executing the function directly if the globalpointer is local**. This means that no remote data at all is needed (and hence computed) if the communicator returns IsDistributed() -> false.
This also means that additional storage can be avoided completely in the serial case.
In the practice this means that we can further optimize the code to have no overhead in the serial case as follows:

~~~cpp
//first of all collect all the neighbours into a list ----> NOT ACTUALLY NEEDED IN SERIAL
GlobalPointerVector< Node<3 > > gp_list;

if(DefaultDataCommunicator->IsDistributed() == true) //an emty gp_list is enough in the serial case
    for(auto& node : model_part.Nodes)
        for(auto& gp : global_pointers_to_neighbours)
             gp_list.push_back(gp);

//now create the pointer communicator --> NOTHING IS ACTUALLY DONE IN THE SERIAL CASE!
PointerCommunicator pointer_comm(DefaultDataCommunicator, gp_list); //here we prepare the communication

//define a function to access to the data ----> NO COMMUNICATION HAPPENS IN THE SERIAL CASE
//we only store the access function within result_proxy and we will use it on every "Get" call
result_proxy = pointer_comm.Apply<double>(
            [](GlobalPointer<Node<3> >& gp) -> double
            {
                  return gp->FastGetSolutionStepValue(TEMPERATURE);
            }
      );  //here all communications happen !!

for(auto& node : model_part.Nodes)
    double Tavg = 0;
    const auto& global_pointers_to_neighbours = node.GetValue(GP_NODAL_NEIGHBOURS);
    for(auto& gp : global_pointers_to_neighbours)
         Tavg += result_proxy.Get(gp); //here we get the result of the application of the function onto gp. This is so BOTH for LOCAL GPs and for REMOTE GPs
    node.SetValue(AVERAGE_TEMPERATURE, Tavg/(global_pointers_to_neighbours.size());
}
~~~
with such modification the code is almost as fast in serial as the original code: a function call by function pointer is done on every global pointer, but the lock that was needed for the weak pointer is avoided.

RiccardoRossi commented:
just to add that "Apply" is templated in the result_type. 
this implies that the function passed as parameter (here i use a lambda function, but it could be anything) 
can return any type of object, and this object **will be transparently returned to the caller object** as long as the object is serializable and does not contain pointers.

RiccardoRossi committed:
small correction to make this to work correctly also for containers which are not nodes

RiccardoRossi committed:
naming change in accordance to coding standards

RiccardoRossi committed:
naming change in accordance to coding standards

RiccardoRossi committed:
making data communicator const

RiccardoRossi committed:
moving coloring to constructor

RiccardoRossi commented:
@jcotela i think i addressed all of your comments.

regarding the SendRecv i would merge this, and once it is in move the templated SendRecv to the data communicator. 
Once that is done i would remove the implementation inside the pointer_communicator and inside the global_pointer_utilities.

other TODOs are related to missing Send and Recv separate functions (no hurry to implement).
Also generic Scatter/gather/broadcast are not present 

as a final comment, i now added PointerComm to the MPI core **however pointer_comm** perfectly works in serial mode with the default serial communicator.
This means that it would probably be better to have it in the includes subdir of the core

pooyan-dadvand commented:
@RiccardoRossi Thanks for your detail explanation!

Now you are passing a constructed gp_list to the constructor. Maybe we can open an alternative way of creating the communicator with a function which creates the list:
~~~cpp
//Create a pointer communicator with function --> NOTHING IS ACTUALLY DONE IN THE SERIAL CASE!
PointerCommunicator pointer_comm(DefaultDataCommunicator, 
          [model_part.Nodes](GlobalPointerVector< Node<3 > >& GpList) {
              for(auto& node : model_part.Nodes)
                  for(auto& gp : global_pointers_to_neighbours)
                       GpList.push_back(gp);
          }
);  //here we prepare the communication

//define a function to access to the data ----> NO COMMUNICATION HAPPENS IN THE SERIAL CASE
//we only store the access function within result_proxy and we will use it on every "Get" call
result_proxy = pointer_comm.Apply<double>(
            [](GlobalPointer<Node<3> >& gp) -> double
            {
                  return gp->FastGetSolutionStepValue(TEMPERATURE);
            }
      );  //here all communications happen !!

for(auto& node : model_part.Nodes)
    double Tavg = 0;
    const auto& global_pointers_to_neighbours = node.GetValue(GP_NODAL_NEIGHBOURS);
    for(auto& gp : global_pointers_to_neighbours)
         Tavg += result_proxy.Get(gp); //here we get the result of the application of the function onto gp. This is so BOTH for LOCAL GPs and for REMOTE GPs
    node.SetValue(AVERAGE_TEMPERATURE, Tavg/(global_pointers_to_neighbours.size());
}
~~~

and having some predefined functions for usual use cases to reduce the code:
~~~cpp
//Create a pointer communicator with function --> NOTHING IS ACTUALLY DONE IN THE SERIAL CASE!
PointerCommunicator pointer_comm(DefaultDataCommunicator, CollectNeighboursGlobalPointers( model_part.Nodes));  //here we prepare the communication

//define a function to access to the data ----> NO COMMUNICATION HAPPENS IN THE SERIAL CASE
//we only store the access function within result_proxy and we will use it on every "Get" call
result_proxy = pointer_comm.Apply<double>(GetVariableFromGlobalPointer(TEMPERATURE));  //here all communications happen !!

for(auto& node : model_part.Nodes)
    double Tavg = 0;
    const auto& global_pointers_to_neighbours = node.GetValue(GP_NODAL_NEIGHBOURS);
    for(auto& gp : global_pointers_to_neighbours)
         Tavg += result_proxy.Get(gp); //here we get the result of the application of the function onto gp. This is so BOTH for LOCAL GPs and for REMOTE GPs
    node.SetValue(AVERAGE_TEMPERATURE, Tavg/(global_pointers_to_neighbours.size());
}
~~~


This way we are hiding the gp_list considering it as implementation detail. 

Nevertheless I should check the implementation to see if is viable

RiccardoRossi commented:
@pooyan-dadvand i have no problem in adding the constructor you propose, i think it would look something like

~~~cpp
GlobalPointerCommunicator(DataCommunicator& rComm, 
                                              std::function< GlobalPointersVector< TPointerDataType >(void)& 
         InputFunction )
        :
        mrDataCommunicator(rComm)
    {
        
         if(mrDataCommunicator.IsDistributed())
        {
             auto gp_list = InputFunction() //note that we do not need to pass the list, we can obtain it from the function
            AddPointers(gp_list.begin(),gp_list.end());

            //compute communication plan
            std::vector<int> send_list;
            send_list.reserve( mNonLocalPointers.size() );
            for(auto& it : mNonLocalPointers)
                send_list.push_back( it.first );

             std::sort(send_list.begin(), send_list.end());
            mColors = MPIColoringUtilities::ComputeCommunicationScheduling(send_list, mrDataCommunicator);
        }
    }
~~~

the difference from your proposal is that i am not passing gp_list to the function, but rather getting it as a return value.

what i don't like about this is (and also about my implementation of the proxy) is that we cannot pass a functor (i mean a class with an operator (...) mimicking the interface).
The only (minor) problem is that the TInputFunctionType should be a template parameter of the constructor if we want to allow that. 
I am actually very willing to add this template parameter both for the new constructor as well as for the apply in order to allow the use of functors.

pooyan-dadvand commented:
> the difference from your proposal is that i am not passing gp_list to the function, but rather getting it as a return value.

:+1: 

> I am actually very willing to add this template parameter both for the new constructor as well as for the apply in order to allow the use of functors.

I think it would be good idea!

RiccardoRossi committed:
minor

RiccardoRossi committed:
moving coloring to the non-mpi core

RiccardoRossi committed:
exporting pointer capabilities

RiccardoRossi committed:
function to get by map

RiccardoRossi committed:
moving to the non mpi core the mpi_coloring_utilities - should solve some linking issues

RiccardoRossi committed:
Merge branch 'master' into pointer_communicator

RiccardoRossi committed:
defining independent variables and making this PR independent on others

philbucher commented on kratos/CMakeLists.txt:
```diff
@@ -100,6 +101,7 @@ set( KRATOS_CORE_SOURCES
     ${CMAKE_CURRENT_SOURCE_DIR}/conditions/mesh_condition.cpp;
     ${CMAKE_CURRENT_SOURCE_DIR}/processes/apply_periodic_boundary_condition_process.cpp;
     ${CMAKE_CURRENT_SOURCE_DIR}/linear_solvers/amgcl_solver_impl.cpp;
+    ${CMAKE_CURRENT_SOURCE_DIR}/utilities/mpi_coloring_utilities.cpp
```
ouch...

now we have an MPI-core, why not putting it there?
If I understand correctly this is only needed for MPI right?

RiccardoRossi replied:
the mpi_coloring_utilities.cpp does not have any dependency on mpi (it only depends on DataCommunicator) and is included in the pointer_communicator which can be used both in serial and in parallel code (so when the mpi core is not loaded)

this means that as a utility it needs to go in the core and in the mpi_core

jcotela replied:
I did not understand what do you mean by 
> this means that as a utility it needs to go in the core and in the mpi_core

Do we need to add it to both libraries?

Also, if it has no relation to mpi, why is it called `mpi_....`?

RiccardoRossi replied:
the phrase should have been:
in the core and NOT in the mpi_core.

just think: pointer_comm can be used when you have no mpi right? it includes the mpi_coloring_utilities.h , so this file needs to be available when you compile the pointer_comm.

maybe we could rename it as 

mpi_coloring_utilities -> communication_coloring_utilities

jcotela replied:
:+1: for the rename.

RiccardoRossi replied:
@jcotela i would prefer to make the renaming in a different PR as soon as this is merged. 

the point is that here i already moved the mpi_coloring_utilities, and i would really like to avoid further messing up the same file.

RiccardoRossi replied:
(note that i am referring to renaming the name of the file)

philbucher commented on kratos/python/add_processes_to_python.cpp:
```diff
@@ -110,6 +111,40 @@ void CalculateEmbeddedVariableFromSkinArray(
     rDistProcess.CalculateEmbeddedVariableFromSkin(rVariable, rEmbeddedVariable);
 }
 
+std::unordered_map<int, std::vector<int> > GetNeighbourIds(
```
I don't think this should be in "add_processes_to_python.cpp"

RiccardoRossi replied:
I see your point, nevertheless i am not really sure of where to put it

RiccardoRossi committed:
adding missing files

RiccardoRossi committed:
adding missing files

RiccardoRossi committed:
adding missing file

RiccardoRossi commented:
the basic idea of the latest changes is to make #4729 independent on this so that we have time to improve the current PR while pushing through #4729 and #4779

RiccardoRossi committed:
adding functor cpsnstrcutor

RiccardoRossi committed:
merging master

RiccardoRossi committed:
accepting functor both in constructor and in apply - also allows to avoid specifying template argument in Apply function

RiccardoRossi committed:
test can now run with or without mpi, depending on the value of --using-mpi

RiccardoRossi committed:
cleanup for global pointer utilities

RiccardoRossi committed:
further simplifying the call so that the template type can be automatically deduced

RiccardoRossi committed:
typo

RiccardoRossi committed:
testing functor also for apply

RiccardoRossi committed:
adding test_neighbours to the suite

RiccardoRossi committed:
adding Update functionality and minor cleanup

roigcarlo commented on kratos/utilities/pointer_communicator.h:
```diff
@@ -0,0 +1,454 @@
+            //         for(unsigned int i=0; i<remote_data.size(); ++i)
+            //             non_local_data[gps_to_be_sent[i]] = remote_data[i];
+            //     }
+            // }
```
Shouldn't we remove this?

RiccardoRossi replied:
please don't, if something goes wrong this is fundamental to debug it, and it takes a while to rewrite it

roigcarlo committed:
Kratos api problems

RiccardoRossi committed:
Merge branch 'master' into pointer_communicator

RiccardoRossi committed:
rmeoving old version

RiccardoRossi committed:
testing with functors

RiccardoRossi committed:
adding the implementation of two functors

RiccardoRossi committed:
removing explicit check on --using-mpi

RiccardoRossi committed:
adding a comment

RiccardoRossi commented:
@philbucher after merging the latest master it does work without explicitly importing mpi.

RiccardoRossi committed:
some polishing

RiccardoRossi committed:
storing a reference to the index list

RiccardoRossi committed:
function moved to process

RiccardoRossi committed:
removing move semantics from functor

RiccardoRossi committed:
simplifying fuynction call

pooyan-dadvand commented:
@KratosMultiphysics/technical-committee approves this as first iteration and comments and suggestions are welcomed!

RiccardoRossi committed:
Merge branch 'master' into pointer_communicator

RiccardoRossi committed:
adapting to new global_pointer_vector behaviour

roigcarlo reviewed (approved):
:+1:
